src/standalone/Snow/neural_snow/ModelTools.jl

```diff
@@ -0,0 +1,323 @@
+module ModelTools
+using Flux, LinearAlgebra
+using Flux.Data: DataLoader
+using DataFrames, Dates
+export make_model,
+    get_model_ps,
+    settimescale!,
+    setoutscale!,
+    LRmodel,
+    make_timeseries,
+    trainmodel!
+
+"""
+    make_model(nfeatures, n, z_idx, p_idx; in_scale, dtype)
+
+Create the neural network to be trained, with initial scaling weights.
+
+# Arguments
+- `nfeatures::Int`: indicates number of features.
+- `n::Int`: the value of the hyperparameter n.
+- `z_idx::Int`: The index of the data vectors pertaining to the depth (z) values.
+- `p_idx::Int`: The index of the data vectors pertaining to the precipitation values.
+-  `in_scale::Vector{<:Real}`: Optional scaling constants for each input feature.
+- `dtype::Type`: Sets type of output model. Default is Float32.
+"""
+function make_model(
+    nfeatures::Int,
+    n::Int,
+    z_idx::Int,
+    p_idx::Int;
+    in_scale::Union{Vector{<:Real}, Nothing} = nothing,
+    dtype::Type = Float32,
+)
+    in_scales =
+        (isnothing(in_scale)) ? Matrix{dtype}(diagm(ones(nfeatures))) :
+        Matrix{dtype}(diagm((1.0 ./ in_scale)))
+    get_relus = Matrix{dtype}([1 0 0; 0 1 0; 1 0 -1])
+    get_min = Matrix{dtype}([1 1 -1; 0 1 0])
+    get_max = Matrix{dtype}([1 -1])
+    model = Chain(
+        pred = SkipConnection(
+            Chain(
+                scale = Dense(in_scales, false, identity),
+                l1 = Dense(nfeatures, n * nfeatures, relu),
+                l2 = Dense(n * nfeatures, nfeatures, elu),
+                l3 = Dense(nfeatures, 1),
+            ),
+            vcat, #returns [predicted value, input...]
+        ),
+        get_boundaries = Parallel(
+            vcat,
+            up_bound = x -> relu.(x[1, :])' .* (x[p_idx + 1, :] .> 0)', # = upper = relu(upper)
+            low_bound = x -> x[z_idx + 1, :]', # = z = relu(z)
+            output_pos = x -> x[1, :]',
+        ),
+        final_scale = Dense(
+            Matrix{dtype}(diagm([1.0, 1.0, 1.0])),
+            false,
+            identity,
+        ),
+        #output_no_thresh = x -> x[3, :]'
+        apply_relus = Dense(get_relus, false, relu),  #returns relu(upper) = upper, relu(z) = z, relu(upper - pred_z)
+        apply_upper = Dense(get_min, false, relu), #returns relu(min(pred_z, upper)+z), relu(z) = z
+        apply_lower = Dense(get_max, false, identity), #returns relu(min(pred_z, upper)+z) - relu(z) = max(min(pred_Z, upper), lower)
+    )
+    return model
+end
+# Do we want to modify this to take a list of input features, and the names of the depth and precipitation variables,
+# instead of specifying nfeatures, z_idx, p_idx? Same functionality but simplifies the usage pipeline.
+
+
+"""
+    get_model_ps(model)
+
+Return the trainable weights for the developed neural model.
+
+# Arguments
+- `model::Chain`: the neural model to be used.
```

Comment:
IIUC, `Chain` is a `Flux` type, the neural model is of type `Chain`.  Can the linear model be expressed as a Chain also, but with fewer layers/fixed width of the layer? Does it need its own type?

Reply:
If so, we wouldnt need two methods for each of these, right? get_model_ps, evaluate, etc.

Reply:
there's no need to get the model parameters for the linear model, as the linear model as an object is just the coefficient/parameter list (it's trained directly as A\b upon creation, calling the model just returns the vector of coefficients), there is an evaluate call already written